app/models/git_repository.rb

```diff
@@ -1,22 +1,23 @@
+require "rugged"
+
 class GitRepository < ApplicationRecord
   include AuthenticationMixin
 
   GIT_REPO_DIRECTORY = Rails.root.join('data/git_repos')
 
-  validates :url, :format => URI::regexp(%w(http https)), :allow_nil => false
-  validate  :check_path
+  validates :url, :format => URI::regexp(%w(http https file)), :allow_nil => false
 
   default_value_for :verify_ssl, OpenSSL::SSL::VERIFY_PEER
   validates :verify_ssl, :inclusion => {:in => [OpenSSL::SSL::VERIFY_NONE, OpenSSL::SSL::VERIFY_PEER]}
 
   has_many :git_branches, :dependent => :destroy
   has_many :git_tags, :dependent => :destroy
-  after_destroy :delete_repo_dir
+  after_destroy :delete_repo_dir # TODO: Need to distribute this to all systems
 
   INFO_KEYS = %w(commit_sha commit_message commit_time name).freeze
 
   def refresh
-    ensure_repo
+    update_repo
```

Comment:
I realize this method existed prior to you making changes, but a comment here (for `def refresh` would probably help explain what is happening.  For example:

```
# Fetches and merges the target branch from origin into the existing branch,
# and updates the remote's tags and branches.
```

I know the method is mostly self explanatory, but `update_repo` (compared to `refresh_branches` and `refresh_tags`) is not since I am unsure what `git` actions are being taken by `update_repo` as part of this `refresh`.  I am bringing this up because I have had to jump around at lot between methods as part of this review, so I would guess it would be confusing to other developers as well.

Reply:
Also, in regards to the specifics of this call in particular:

Is `handling_worktree_errors` going to catch something like a dirty git state when trying to do a `fetch_and_merge`?  Also, does it do a `git merge`, or a `rebase`?  I feel like a rebase is better in our use case so extra merge commits aren't being created (since we should never be creating new unique commits in our use cases).

Reply:
fetch_and_merge is basically a pull/rebase but it's one of our own methods, not a method from rugged.  That aspect hasn't changed in this PR, and I'm less interested in fixing that specific thing here, since it works as is.

Most of the changes in the first commit are around getting a handle to the underlying repo without actually updating it.  Previously, when you got a handle, it automatically did the pull, but I had cases where I *don't* want to do the pull right away, or more specifically I want control on when to do the pull.

Reply:
I just realized I didn't answer your question...

handling_worktree_errors mostly just handles network connectivity issues.  You can't have a dirty git state because it's a bare repo under the covers